File: 4_Lookback Option/Option_Lookback_MonteCarlo.py
from math import log, exp, sqrt
import numpy as np
import time
import multiprocessing
# payoff = max(Smax,τ − Sτ , 0)

def lookback_MC(StMax, St, time_left_to_maturity, r, q, sigma, n, sims, rep):
    dt = time_left_to_maturity/n
    times = 0
    means = []
    means = np.ndarray(shape = (rep))
    while times < rep:
        optionValues = np.ndarray(shape = (sims))

        for sim in range(sims):
            stockPrices = np.ndarray(shape = (n+1))
            stockPrices[0] = log(St)
            for i in range(n):
                dlnS = np.random.normal(loc = (r-q-sigma**2/2)*dt, scale = sigma*sqrt(dt))
                stockPrices[i+1] = dlnS
            
            stockPrices = np.cumsum(stockPrices)
            stockPrices = np.exp(stockPrices)
            
            if StMax == St:
                maxPrice = max(stockPrices)
            else:
                maxPrice_after_t = max(stockPrices)
                maxPrice = max(StMax, maxPrice_after_t)

            putValue = max(maxPrice - stockPrices[-1], 0) * exp(-r * time_left_to_maturity)
            optionValues[sim] = putValue

        mean = np.mean(optionValues)
        means[times] = mean
        times += 1

    sdOfRep = np.std(means)
    meanOfRep = np.mean(means)
    upperBound = meanOfRep + 2*sdOfRep
    upperBound = round(upperBound, 6)
    lowerBound = meanOfRep - 2*sdOfRep
    lowerBound = round(lowerBound, 6)
    bounds = [lowerBound, upperBound]
    print("==================================================")
    print(f"Lookback Option : European Put")
    print(f'[ Smax,t = {StMax} ]')
    print("--------------------------------------------------")
    print(f"平均 : {round(meanOfRep, 6)}")
    print(f"標準誤 : {round(sdOfRep, 6)}")
    print(f"九十五趴信賴區間 : {bounds}")

    return meanOfRep




# main
St = 50
T = 0.25
r = 0.1
q = 0
sigma = 0.4
n = 100
sims = 10000
reps = 20

# StMax = 50
# lookback_MC(StMax, St, T, r, q, sigma, n, sims, reps)
# StMax = 60
# lookback_MC(StMax, St, T, r, q, sigma, n, sims, reps)
# StMax = 70
# lookback_MC(StMax, St, T, r, q, sigma, n, sims, reps)

start = time.perf_counter()

if __name__ == '__main__':
    StMax = 50
    p1 = multiprocessing.Process(target = lookback_MC, args = [StMax, St, T, r, q, sigma, n, sims, reps])
    StMax = 60
    p2 = multiprocessing.Process(target = lookback_MC, args = [StMax, St, T, r, q, sigma, n, sims, reps])
    StMax = 70
    p3 = multiprocessing.Process(target = lookback_MC, args = [StMax, St, T, r, q, sigma, n, sims, reps])

    p1.start()
    p2.start()
    p3.start()
    p1.join()
    p2.join()
    p3.join()
    finish = time.perf_counter()
    print(f'Process finished in {round(finish - start, 2)} second(s).')

File: 4_Lookback Option/Option_Lookback_MonteCarlo.cpp
# include <iostream>
# include <cmath>
# include <vector>
# include <random>
# include <chrono>


using namespace std;
// lookback option with floating strike
// call payoff = max(Sτ - Smin,τ, 0)
// put payoff = max(Smax,τ − Sτ , 0)
vector<double> lookback_MC_call(double StMin, double St, double time_left_to_maturity, double r, double q, double sigma, int n, int sims, int rep){
    double dt = time_left_to_maturity/n;
    int times = 0;
    
    default_random_engine generator;
    generator.seed(chrono::system_clock::now().time_since_epoch().count());
    normal_distribution<double> distribution((r - q - 0.5*(pow(sigma, 2)))*dt, sigma*sqrt(dt));

    vector<double> means;
    while(times < rep){
        vector<double> optionValue;

        for(int i = 0; i < sims; i++){
            double minPrice = StMin;
            vector<double> stockPrices;
            stockPrices.push_back(log(St));

            for(int j = 1; j < n+1; j++){
                double dlnS = distribution(generator);
                double sample = stockPrices[j-1] + dlnS;
                stockPrices.push_back(sample);
            }
            for(int k = 0; k < n+1; k++){
                stockPrices[k] = exp(stockPrices[k]);
                if(stockPrices[k] < minPrice){
                    minPrice = stockPrices[k];
                }
            }
            // for each loop, single price path is simulated...
            double callValue = max(stockPrices[n] - minPrice, 0.0)*exp(-r * time_left_to_maturity);
            // cout << putValue << endl;
            optionValue.push_back(callValue);
            }
        
        double sum = 0;
        for(int i = 0; i < sims; i++){
            sum += optionValue[i];
        }
        double mean = sum/sims;
        means.push_back(mean);
        times += 1;
        }
    
    double sum_means = 0;
    for(int l = 0; l < rep; l++){
        sum_means += means[l];
    }
    double meanOfRep = sum_means/rep;
    
    double var = 0.0;
    for(int n = 0; n < rep; n++){
        var += (means[n] - meanOfRep) * (means[n] - meanOfRep);
    }
    var = var/rep;
    double sdOfRep = sqrt(var);
    double upper = meanOfRep + 2*sdOfRep;
    double lower = meanOfRep - 2*sdOfRep;
    vector<double> results = {meanOfRep, lower, upper};
    cout << "==================================================" << endl;
    cout << "Lookback Option : European call " << endl;
    cout << "[ Smin,t = "<< StMin << " ]" << endl;
    cout << "--------------------------------------------------" << endl;
    cout << "mean : " << meanOfRep << endl;
    cout << "standard error : " << sdOfRep << endl;
    cout << "0.95 confidence interval : [ " << lower << ", " << upper << " ]" << endl;
    cout << endl;
    return results;
}

vector<double> lookback_MC_put(double StMax, double St, double time_left_to_maturity, double r, double q, double sigma, int n, int sims, int rep){
    double dt = time_left_to_maturity/n;
    int times = 0;
    
    default_random_engine generator;
    generator.seed(chrono::system_clock::now().time_since_epoch().count());
    normal_distribution<double> distribution((r - q - 0.5*(pow(sigma, 2)))*dt, sigma*sqrt(dt));

    vector<double> means;
    while(times < rep){
        vector<double> optionValue;

        for(int i = 0; i < sims; i++){
            double maxPrice = StMax;
            vector<double> stockPrices;
            stockPrices.push_back(log(St));

            for(int j = 1; j < n+1; j++){
                double dlnS = distribution(generator);
                double sample = stockPrices[j-1] + dlnS;
                stockPrices.push_back(sample);
            }
            for(int k = 0; k < n+1; k++){
                stockPrices[k] = exp(stockPrices[k]);
                if(stockPrices[k] > maxPrice){
                    maxPrice = stockPrices[k];
                }
            }
            // for each loop, single price path is simulated...
            double putValue = max(maxPrice - stockPrices[n], 0.0)*exp(-r * time_left_to_maturity);
            // cout << putValue << endl;
            optionValue.push_back(putValue);
            }
        
        double sum = 0;
        for(int i = 0; i < sims; i++){
            sum += optionValue[i];
        }
        double mean = sum/sims;
        means.push_back(mean);
        times += 1;
        }
    
    double sum_means = 0;
    for(int l = 0; l < rep; l++){
        sum_means += means[l];
    }
    double meanOfRep = sum_means/rep;
    
    double var = 0.0;
    for(int n = 0; n < rep; n++){
        var += (means[n] - meanOfRep) * (means[n] - meanOfRep);
    }
    var = var/rep;
    double sdOfRep = sqrt(var);
    double upper = meanOfRep + 2*sdOfRep;
    double lower = meanOfRep - 2*sdOfRep;
    vector<double> results = {meanOfRep, lower, upper};
    cout << "==================================================" << endl;
    cout << "Lookback Option : European put " << endl;
    cout << "[ Smax,t = "<< StMax << " ]" << endl;
    cout << "--------------------------------------------------" << endl;
    cout << "mean : " << meanOfRep << endl;
    cout << "standard error : " << sdOfRep << endl;
    cout << "0.95 confidence interval : [ " << lower << ", " << upper << " ]" << endl;
    cout << endl;
    return results;
}



int main(){
    double St = 50;
    double T = 0.25;
    double r = 0.1;
    double q = 0;
    double sigma = 0.4;
    int n = 100;
    int sims = 10000;
    int rep = 100;

    lookback_MC_call(30, St, T, r, q, sigma, n, sims, rep);
    lookback_MC_call(40, St, T, r, q, sigma, n, sims, rep);
    lookback_MC_call(50, St, T, r, q, sigma, n, sims, rep);
    
    lookback_MC_put(50, St, T, r, q, sigma, n, sims, rep);
    lookback_MC_put(60, St, T, r, q, sigma, n, sims, rep);
    lookback_MC_put(70, St, T, r, q, sigma, n, sims, rep);
}